src/tools/dhcpv6-discover.c

```diff
@@ -0,0 +1,822 @@
+ *
+ * This function waits for ICMPv6 Router Advertisement messages on the specified
+ * file descriptor until the specified timeout is reached. It processes each
+ * received message and counts the number of valid responses.
+ *
+ * @param fd The file descriptor to read from.
+ * @param tgt The target sockaddr_in6 structure containing the expected source address.
+ * @param ifname The name of the network interface.
+ * @return The number of valid Router Advertisement responses received, or -1 on error.
+ */
+static ssize_t recv_adv(int fd, const struct sockaddr_in6 *tgt, const char *ifname, const unsigned int timeout)
+{
+	struct timespec end = { 0 };
+	unsigned responses = 0;
+
+	// Get the current time and add the timeout
+	clock_gettime(CLOCK_MONOTONIC, &end);
+	end.tv_sec += timeout;
+
+	// Receiving packets until timeout
+	while(true)
+	{
+		// Wait for reply until timeout
+		ssize_t val = 0;
+
+		struct timespec now = { 0 };
+		clock_gettime(CLOCK_MONOTONIC, &now);
+		if(end.tv_sec >= now.tv_sec)
+		{
+			// Calculate the remaining time
+			val = (end.tv_sec - now.tv_sec) * 1000 + (int)((end.tv_nsec - now.tv_nsec) / 1000000);
+			if (val <= 0) // Timeout
+				return responses;
+		}
+
+		// Wait for reply (retries on EINTR)
+		struct pollfd pollfd = { .fd = fd, .events = POLLIN, .revents = 0 };
+		do {
+			val = poll(&pollfd, 1, val);
+		} while (val == -1 && errno == EINTR);
+
+		// Check for errors, logging happens in the calling function
+		if(val < 0)
+			break;
+
+		// Check for timeout
+		if(val == 0)
+			return responses;
+
+		// Received a packet
+		uint8_t buf[1460];
+		struct sockaddr_in6 addr = { 0 };
+		val = recvfromLL(fd, &buf, sizeof(buf), MSG_DONTWAIT, &addr);
+		if (val == -1)
+		{
+			// Ignore EAGAIN as we can retry
+			if (errno != EAGAIN)
+				printf("Error while receiving Router Advertisements on %s: %s\n",
+				       ifname, strerror(errno));
+			continue;
+		}
+
+		// Check that the response came through the right interface
+		if (addr.sin6_scope_id && (addr.sin6_scope_id != tgt->sin6_scope_id))
+			continue;
+
+		// Print the received packet's size and the source address
+		char str[INET6_ADDRSTRLEN] = { 0 };
+		inet_ntop(AF_INET6, &addr.sin6_addr, str,sizeof (str));
+		start_lock();
+		printf("* Received %zd bytes from %s @ %s\n", val, str, ifname);
+
+		// Parse the Router Advertisement
+		if(parse_ra(buf, val) == 0)
+			responses++;
+		end_lock();
+	}
+
+	return -1;
+}
+
+/**
+ * @brief Sends a Router Solicitation message and waits for a Router Advertisement response.
+ *
+ * This function performs Neighbor Discovery (ND) by sending a Router Solicitation (RS)
+ * message to the specified target and waits for a Router Advertisement (RA) response.
+ *
+ * @param fd The file descriptor of the socket to use for sending and receiving messages.
+ * @param ifname The name of the network interface to use for sending the message.
+ *
+ * @return 0 on success, -1 on error, -2 if no response is received after retries.
```

Comment:
I don't see where it returns `-2`

Reply:
I should probably write documentation only when the code is finished...

```suggestion
 * @return 0 on success, -1 on error.
```